cloudevents/work/agent/client/manifestwork.go

```diff
@@ -0,0 +1,195 @@
+			Type:    ManifestsDeleted,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ManifestsDeleted",
+			Message: fmt.Sprintf("The manifests are deleted from the cluster %s", manifestWork.Namespace),
+		})
+
+		eventDataType, err := types.ParseCloudEventsDataType(manifestWork.Annotations[codec.CloudEventsDataTypeAnnotationKey])
+		if err != nil {
+			return nil, err
+		}
+
+		eventType := types.CloudEventsType{
+			CloudEventsDataType: *eventDataType,
+			SubResource:         types.SubResourceStatus,
+			Action:              DeleteRequestAction,
+		}
+
+		if err := c.cloudEventsClient.Publish(ctx, eventType, manifestWork); err != nil {
+			return nil, err
+		}
+
+		c.watcher.Receive(watch.Event{Type: watch.Deleted, Object: manifestWork})
+
+		return manifestWork, nil
+	}
+
+	resourceVersion, err := strconv.ParseInt(manifestWork.ResourceVersion, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the resourceVersion of the manifestwork %s, %v", manifestWork.Name, err)
+	}
+
+	lastResourceVersion, err := strconv.ParseInt(lastWork.ResourceVersion, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the resourceVersion of the manifestwork %s, %v", lastWork.Name, err)
+	}
+
+	if resourceVersion <= lastResourceVersion {
+		klog.Infof("The work %s resource version is less than or equal to cached, ignore", manifestWork.Name)
+		return manifestWork, nil
+	}
+
+	updatedWork := manifestWork.DeepCopy()
+
+	// restore the fields that are maintained by local agent
+	updatedWork.Labels = lastWork.Labels
+	updatedWork.Annotations = lastWork.Annotations
+	updatedWork.Finalizers = lastWork.Finalizers
+	updatedWork.Status = lastWork.Status
+
+	c.watcher.Receive(watch.Event{
+		Type:   watch.Modified,
+		Object: updatedWork,
+	})
+
+	return updatedWork, nil
+}
+
+func (c *ManifestWorksAgentClient) UpdateStatus(ctx context.Context, manifestWork *workv1.ManifestWork, opts metav1.UpdateOptions) (*workv1.ManifestWork, error) {
+	klog.V(4).Infof("updating manifestwork %s status", manifestWork.Name)
+
+	_, err := c.lister.Get(manifestWork.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	eventDataType, err := types.ParseCloudEventsDataType(manifestWork.Annotations[codec.CloudEventsDataTypeAnnotationKey])
+	if err != nil {
+		return nil, err
+	}
+
+	updatedWork := manifestWork.DeepCopy()
+
+	eventType := types.CloudEventsType{
+		CloudEventsDataType: *eventDataType,
+		SubResource:         types.SubResourceStatus,
+		Action:              UpdateRequestAction,
+	}
+
+	if err := c.cloudEventsClient.Publish(ctx, eventType, updatedWork); err != nil {
+		return nil, err
+	}
+	return updatedWork, nil
+}
+
+func (c *ManifestWorksAgentClient) Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error {
+	klog.V(4).Infof("deleting manifestwork %s", name)
+	work, err := c.lister.Get(name)
+	if err != nil {
+		return err
+	}
+
+	c.watcher.Receive(watch.Event{Type: watch.Modified, Object: work})
```

Comment:
what is the intension of this? and I do not think agent should have the permission to delete?

Reply:
good catch, we don't need the delete function for agent, we just need mark the manifest has the deletiontimestamp, this should be finished by update function